### PR TITLE
Improvements for AssetDetails, Tags & Unique Identifiers + Adhoc Reports

### PR DESCRIPTION
**Pull request by fruechel**
Tags & Unique Identifiers are now supported similar to the Ruby client. Adhoc reports can be generated in more than XML.

## Description
See commit messages for more details. They outline each change.

## Motivation and Context
These changes are needed for our use case of Nexpose. All changes are backwards-compatible improvements which is why they are in a PR.

## How Has This Been Tested?
No additional tests have been added but all changes have been verified and are being used by our code. All old tests still work (which is expected given that all changes are backwards-compatible). Changes have only been tested in Python 3. The only potential problem is the new CSV parsing function which has been explicitly tested & verified for Python 2 as well.

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] I have updated the documentation accordingly (if changes are required).
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

There's no docs for any of the areas I touched so I didn't add any either. Same goes for tests.

### CODE REVIEW

File: nexpose/nexpose.py [hunk 1]
```diff
@@ -2032,22 +2043,45 @@ def DownloadReport(self, report_or_id, callback_function=None, block_size=DEFAUL
...
-        assert data[2] == 'Content-Transfer-Encoding: base64'
-        assert data[3] == ''
-        assert data[0] == data[-1][:-2]
+        boundary_top = data[0]
+        content_type = data[1]
+        encoding = data[2]
         body = ''.join(data[4:-1])
+        boundary_bottom = data[-1]
+        if boundary_top != boundary_bottom[:-2]:
+            raise ValueError("Invalid boundary")
```
Comment (fruechel) [inline nexpose/nexpose.py:2061]: I replaced the assertions as per Python recommendation: asserts could be compiled out so throwing exceptions is the preferred method here.

**Comment by gschneider-r7** [comment]
I fixed (well, more like bandaid) the flake8 issues on master so if you rebase travis-ci should pass on this PR.

**Comment by gschneider-r7** [comment]
Also as far as reporting goes, let me know if you have any feedback on #37

**Comment by fruechel** [comment]
I merged in master, seems this worked. As for #37 it looks like we should combine the changes since they touch similar parts. I'll have a closer look. If you want I can remove the report commit from this PR and we work on #37 to incorporate both changes. Does that work for you?

**Comment by gschneider-r7** [comment]
I don't mind having your report changes here as well since this PR could probably get into a released version sooner than #37. Feel free to PR any suggested changes to the `report_configs` branch, then I can review and merge to #37.

**Comment by fruechel** [comment]
Does this mean we can merge these changes into master or do you need anything from me before making that merge?

File: nexpose/nexpose.py [hunk 2]
```diff
@@ -620,15 +622,20 @@ def RequestReportDelete(self, report_id, reportconfiguration_id=0):
...
             return self.ExecuteBasicOnReport("ReportDeleteRequest", report_id)
 
-    def RequestReportAdhocGenerate(self, id):
+    def RequestReportAdhocGenerate(self, id, format='raw-xml-v2', template_id='audit-report'):
         request = """
-<AdhocReportConfig format="raw-xml-v2" template-id="audit-report">
+<AdhocReportConfig format="{format}" template-id="{template_id}">
 <Filters>
-<filter type="scan" id="{0}" />
+<filter type="scan" id="{scan_id}" />
```
Comment (gschneider-r7) [inline nexpose/nexpose.py:629]: where does `scan_id` come from?

Comment (fruechel) [reply nexpose/nexpose.py:629]: It comes from `id` as part of the method signature. It's always been there, I just renamed it for clarity.

**Review by gschneider-r7 — APPROVED** [review]

**Comment by gschneider-r7** [comment]
These changes are published in v0.1.7, available on pypi now.

**Comment by fruechel** [comment]
That was really fast, thank you!

**Pull request merged**